src/main/java/org/openlmis/integration/dhis2/web/period/PeriodMappingController.java

```diff
@@ -0,0 +1,188 @@
+  }
+
+  /**
+   * Retrieves all period mappings. Note that an empty collection rather than a 404 should be
+   * returned if no period mappings exist.
+   */
+  @GetMapping
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<PeriodMappingDto> getAllPeriodMappings(@PathVariable("serverId") UUID serverId,
+                                                     Pageable pageable) {
+    Server server = serverRepository.findById(serverId)
+            .orElseThrow(() -> new NotFoundException(MessageKeys.ERROR_SERVER_NOT_FOUND));
+
+    List<PeriodMapping> periodMappings = server.getPeriodMappingList();
+    List<PeriodMappingDto> periodMappingDtos = periodMappings
+            .stream()
+            .map(PeriodMappingDto::newInstance)
+            .collect(Collectors.toList());
+
+    return Pagination.getPage(periodMappingDtos, pageable);
+  }
+
+  /**
+   * Allows the creation of a new period mapping. If the id is specified, it will be ignored.
+   */
+  @PostMapping
+  @ResponseStatus(HttpStatus.CREATED)
+  @ResponseBody
+  public PeriodMappingDto createPeriodMapping(@PathVariable("serverId") UUID serverId,
+                                              @RequestBody PeriodMappingDto periodMappingDto) {
+    LOGGER.debug("Creating new period mapping");
+    PeriodMapping newPeriodMapping = PeriodMapping.newInstance(periodMappingDto);
+
+    Server server = serverRepository.findById(serverId)
+            .orElseThrow(() -> new NotFoundException(MessageKeys.ERROR_SERVER_NOT_FOUND));
+    newPeriodMapping.setServer(server);
+
+    newPeriodMapping.setId(null);
+    newPeriodMapping = periodMappingRepository.saveAndFlush(newPeriodMapping);
+
+    return PeriodMappingDto.newInstance(newPeriodMapping);
+  }
+
+  /**
+   * Updates the specified period mapping.
+   */
+  @PutMapping(value = "/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public PeriodMappingDto updatePeriodMapping(@PathVariable("id") UUID id,
+                                              @RequestBody PeriodMappingDto periodMappingDto) {
+    if (null != periodMappingDto.getId() && !Objects.equals(periodMappingDto.getId(), id)) {
+      throw new ValidationMessageException(MessageKeys.ERROR_PERIOD_MAPPING_ID_MISMATCH);
+    }
+
+    LOGGER.debug("Updating period mapping");
+    PeriodMapping periodMappingToSave = periodMappingRepository.findById(id).map(periodMapping -> {
+      periodMapping.updateFrom(periodMappingDto);
+      return periodMapping;
+    }).orElseThrow(() -> new NotFoundException(MessageKeys.ERROR_PERIOD_MAPPING_NOT_FOUND));
+
+    periodMappingRepository.saveAndFlush(periodMappingToSave);
+    return PeriodMappingDto.newInstance(periodMappingToSave);
+  }
+
+  /**
+   * Deletes the specified period mapping.
+   */
+  @DeleteMapping(value = "/{id}")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void deletePeriodMapping(@PathVariable("id") UUID id) {
+    if (!periodMappingRepository.existsById(id)) {
```

Comment:
i do not see any logic to delete period mapping

Reply:
added deleting PeriodMapping logic